server/src/main/java/org/opensearch/search/internal/InternalSearchResponse.java

```diff
@@ -78,7 +94,8 @@ public InternalSearchResponse(StreamInput in) throws IOException {
             in.readBoolean(),
             in.readOptionalBoolean(),
             in.readOptionalWriteable(SearchProfileShardResults::new),
-            in.readVInt()
+            in.readVInt(),
+            in.readNamedWriteableList(SearchExtBuilder.class)
```

Comment:
@reta 

I wrote a unit test where I set the version on the StreamInput to be V_3_0_0 (current) and the version on the StreamOutput to be V_2_9_0.  This caused the test to fail because of this logic (not related to my change):

https://github.com/opensearch-project/OpenSearch/blob/f2ca769143d1c589806d5f9adc65e0545eedd753/server/src/main/java/org/opensearch/search/profile/ProfileResult.java#L130-L161

So, basically, those optional Longs don't get written out to the output stream because the writer is on the older version, but the reader is expecting optional Longs because it is >= V_2_10_0.

Can this scenario really happen?  

Are all four scenarios - (reader version, writer version) = (old, old), (old, new), (new, old), (new, new) - possible during a rolling migration?  My guess is yes, but just wanted to confirm.

Also, on second thought, I am not sure that we really need the "if" condition on write, since in all the older versions of the code, this new stuff won't be written to the output stream.

Furthermore, the reader does not know the version of the writer so reading the new stuff based on the version of the reader seems flawed.  Shouldn't the writer put its version in the output stream so the reader can use this information to know what to expect in the incoming stream?

If (r_v, w_v) = (2.10, 2.9) is possible, the code above breaks.

Reply:
https://github.com/opensearch-project/OpenSearch/blob/301c766075a4a91870439482cab4cfb09d9299bd/libs/core/src/main/java/org/opensearch/core/common/io/stream/StreamInput.java#L120-L125

I think this answered my question.

Reply:
@austintlee sorry, a bit late, the bwc tests are automatic, including mixing cluster scenarios, so in theory we could have V3.0.0, V2.10.0 and V2.9.0 running at the same time. Your change could go only as far as 3.0.0 and 2.10.0 (after backport)

Reply:
> So, basically, those optional Longs don't get written out to the output stream because the writer is on the older version, but the reader is expecting optional Longs because it is >= V_2_10_0.

I haven't see you test but it looks to me in / out have wrong versions. For example, if coordinator node is 3.0.0, it will be consulting the version of the node to write response to (2.9.0 or 3.0.0), consequently the  writing node would consult the version of the node it is replying (2.9.0 or 3.0.0).